tiledb/common/evaluator/evaluator.h

```diff
@@ -0,0 +1,411 @@
+/**
+ * @file   tiledb/common/evaluator/evaluator.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2024 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines a function evaluator class and a set of caching policies
+ * that can be used to configure the evaluator's behavior.
+ */
+
+#ifndef TILEDB_COMMON_EVALUATOR_H
+#define TILEDB_COMMON_EVALUATOR_H
+
+#include <future>
+#include <list>
+
+namespace tiledb::common {
+
+/**
+ * Abstract class for caching policies.
+ * This contains most of the logic of the caching mechanism
+ * except budgeting which is policy specific.
+ *
+ * @tparam Key The type of key
+ * @tparam Value The type of value
+ */
```

Comment:
The class invariants of the three containers should be documented here in the class header. Because they interact with each other, documenting them one at a time is insufficient to fully describe the class behavior.